llvm/include/llvm/ADT/PagedVector.h

```diff
@@ -0,0 +1,301 @@
+      auto Copy = *this;
+      ++*this;
+      return Copy;
+    }
+
+    std::ptrdiff_t operator-(MaterialisedIterator const &Other) const {
+      assert(PV == Other.PV);
+      // If they are on the same table we can just subtract the indices.
+      // Otherwise we have to iterate over the pages to find the difference.
+      // If a page is invalid, we skip it.
+      if (PV == Other.PV) {
+        return ElementIdx - Other.ElementIdx;
+      }
+
+      auto ElementMin = std::min(ElementIdx, Other.ElementIdx);
+      auto ElementMax = std::max(ElementIdx, Other.ElementIdx);
+      auto PageMin = ElementMin / PAGE_SIZE;
+      auto PageMax = ElementMax / PAGE_SIZE;
+
+      auto Count = 0ULL;
+      for (auto PageIdx = PageMin; PageIdx < PageMax; ++PageIdx) {
+        if (PV->PageToDataIdx[PageIdx] == invalidPage()) {
+          continue;
+        }
+        Count += PAGE_SIZE;
+      }
+      Count += ElementMax % PAGE_SIZE;
+      Count += PAGE_SIZE - ElementMin % PAGE_SIZE;
+
+      return Count;
+    }
+
+    // When dereferencing the iterator, we materialise the page if needed.
+    T const &operator*() const {
+      assert(ElementIdx < PV->Size);
+      assert(PV->PageToDataIdx[ElementIdx / PAGE_SIZE] != invalidPage());
+      return *((ElementIdx % PAGE_SIZE) +
+               PV->PageToDataIdx[ElementIdx / PAGE_SIZE]);
+    }
+
+    // Equality operator.
+    bool operator==(MaterialisedIterator const &Other) const {
+      // Iterators of two different vectors are never equal.
+      if (PV != Other.PV) {
+        return false;
+      }
+      // Any iterator for an empty vector is equal to any other iterator.
+      if (PV->empty()) {
+        return true;
+      }
+      // Get the pages of the two iterators. If between the two pages there
+      // are no valid pages, we can condider the iterators equal.
+      auto PageMin = std::min(ElementIdx, Other.ElementIdx) / PAGE_SIZE;
+      auto PageMax = std::max(ElementIdx, Other.ElementIdx) / PAGE_SIZE;
+      // If the two pages are past the end, the iterators are equal.
+      if (PageMin >= PV->PageToDataIdx.size()) {
+        return true;
+      }
+      // If only the last page is past the end, the iterators are equal if
+      // all the pages up to the end are invalid.
+      if (PageMax >= PV->PageToDataIdx.size()) {
+        for (auto PageIdx = PageMin; PageIdx < PV->PageToDataIdx.size();
+             ++PageIdx) {
+          if (PV->PageToDataIdx[PageIdx] != invalidPage()) {
+            return false;
+          }
+        }
+        return true;
+      }
+
+      auto *Page1 = PV->PageToDataIdx[PageMin];
+      auto *Page2 = PV->PageToDataIdx[PageMax];
+      if (Page1 == invalidPage() && Page2 == invalidPage()) {
+        return true;
+      }
+      // If the two pages are the same, the iterators are equal if they point
+      // to the same element.
+      if (PageMin == PageMax) {
+        return ElementIdx == Other.ElementIdx;
+      }
```

Comment:
drop `{}` on single-line blocks (here and elsewhere) per LLVM style

Reply:
Ok, I think I have done it everywhere. Resolving.